localstack/services/firehose/provider.py

```
@@ -0,0 +1,610 @@
+                record["kinesisRecordMetadata"] = {
+                    "shardId": "shardId-000000000000",
+                    # not really documented what AWS is using internally - simply using a random UUID here
+                    "partitionKey": str(uuid.uuid4()),
+                    "approximateArrivalTimestamp": timestamp(
+                        float(_get_entry(record, "ApproximateArrivalTimestamp")) / 1000,
+                        format=TIMESTAMP_FORMAT_MICROS,
+                    ),
+                    "sequenceNumber": next_sequence_number(),
+                    "subsequenceNumber": "",
+                }
+
+    def _preprocess_records(self, processor: Dict, records: List[Record]) -> List[Dict]:
+        """Preprocess the list of records by calling the given processor (e.g., Lamnda function)."""
+        proc_type = processor.get("Type")
+        parameters = processor.get("Parameters", [])
+        parameters = {p["ParameterName"]: p["ParameterValue"] for p in parameters}
+        if proc_type == "Lambda":
+            lambda_arn = parameters.get("LambdaArn")
+            # TODO: add support for other parameters, e.g., NumberOfRetries, BufferSizeInMBs, BufferIntervalInSeconds, ...
+            client = aws_stack.connect_to_service("lambda")
+            records = keys_to_lower(records)
+            # Convert the record data to string (for json serialization)
+            for record in records:
+                if "data" in record:
+                    record["data"] = to_str(record["data"])
+                if "Data" in record:
+                    record["Data"] = to_str(record["Data"])
+            event = {"records": records}
+            event = to_bytes(json.dumps(event))
+            response = client.invoke(FunctionName=lambda_arn, Payload=event)
+            result = response.get("Payload").read()
+            result = json.loads(to_str(result))
+            records = result.get("records", []) if result else []
```

Does this mean that if the Lambda function errors out or returns some invalid response that records we attempted to process are silently dropped?

---

Yes, that's correct. However, I am not sure how the AWS service handles failures of processing lambdas (I just migrated the behavior as it was before).

---

I'm not sure either actually, I've never had to use the Lambda integration 🤔 I'm guessing the records just get put into S3 like in other failure cases with Firehose. Might want to add a warning here.